### PR TITLE
feat: add Rope data structure

### PR DESCRIPTION
**Pull request by hieudo-dev**
## Description
This pull request introduces a basic implementation of the Rope data structure.

## Rope
A Rope is a data structure designed for efficient manipulation of large strings of text by dividing the text into smaller segments represented as nodes in a binary tree. This structure optimizes operations like insertion, deletion, and concatenation by minimizing the need for extensive copying and rearranging of characters. Ropes are particularly useful for applications dealing with very large texts, such as text editors or DNA sequencing, where traditional string manipulation techniques may become inefficient due to performance constraints.

### Supported operations
- Concat
- Split
- Insert
- Delete
- Slice

### Sources and References
[Wikipedia](https://en.wikipedia.org/wiki/Rope_(data_structure))
[Rope Implementation Overview](https://www.boost.org/sgi/stl/ropeimpl.html)

### CODE REVIEW

File: src/data_structures/rope.rs [hunk 1]
```diff
@@ -37,151 +73,194 @@ impl Display for Rope {
...
-                            Some(Box::new(Rope::Node {
-                                left,
-                                right: Some(right),
-                                weight,
-                            })),
-                            None,
-                        ),
-                        (_, _) => panic!("Invalid split results"),
+                    let (left, right) = right.split_at(index - node.weight);
+                    // println!("greater {}", right.asunwrap());
```
Comment (alexfertel) [inline src/data_structures/rope.rs]: 👀

File: src/data_structures/rope.rs [hunk 2]
```diff
@@ -37,151 +73,194 @@ impl Display for Rope {
...
         };
-        let (_, right) = remain.split(length);
-        let weight = left.as_ref().unwrap().get_weight();
-        return Box::new(Rope::Node {
-            left,
-            right,
-            weight,
-        });
+
+        println!("{} {}", remain.as_ref().to_string(), length);
```
Comment (alexfertel) [inline src/data_structures/rope.rs]: 👀

File: src/data_structures/rope.rs [hunk 3]
```diff
@@ -228,94 +307,152 @@ mod tests {
...
+                weight: 4,
+            }),
+            right: leaf!(String::from("baz")),
+            weight: 4,
+        };
+        let result = rope.delete_at(0, 4);
+        match result {
+            Err(e) => panic!("Unexpected error: {}", e),
+            Ok(rope) => assert_eq!(rope.to_string(), "baz"),
+        }
```
Comment (alexfertel) [inline src/data_structures/rope.rs]:
```suggestion
        let result = rope.delete_at(0, 4).unwrap();
        assert_eq!(rope.to_string(), "baz");
```

File: src/data_structures/rope.rs [hunk 4]
```diff
@@ -228,94 +307,152 @@ mod tests {
...
-        });
+        };
 
-        rope = rope.delete(4, 5);
-        assert_eq!(rope.to_string(), "foo baz");
+        let result = rope.delete_at(4, 3);
+        match result {
+            Err(e) => panic!("Unexpected error: {}", e),
+            Ok(rope) => assert_eq!(rope.to_string(), "foo "),
+        }
```
Comment (alexfertel) [inline src/data_structures/rope.rs]:
```suggestion
        let result = rope.delete_at(4, 3).unwrap();
        assert_eq!(rope.to_string(), "foo ");
```

**Review by alexfertel — COMMENTED** [review]
Looks good! Only a few small comments left and it's good to go!

File: src/data_structures/rope.rs (hunk 1 above)
Comment (hieudo-dev) [reply src/data_structures/rope.rs]: oops

**Review by alexfertel — APPROVED** [review]

**Pull request merged**